### PR TITLE
fix(projects): ignore terminal projects in shortname dedup

### LINKED ISSUE #553: Project rename blocked by cancelled/completed projects in shortname dedup
## Bug

When renaming a project via `PATCH /api/projects/:id`, the shortname deduplication logic considers **all** projects in the company — including cancelled and completed ones. This silently prevents renaming a project to a name that was previously used by a now-cancelled project.

### Steps to reproduce

1. Create a project named "Growth Research"
2. Cancel it
3. Create another project (which auto-names to "Growth Research 2")
4. Try to rename "Growth Research 2" → "Growth Research" via PATCH
5. The API returns 200 but the name stays "Growth Research 2" — the dedup silently rewrites it back

### Expected behavior

Cancelled and completed projects should not block name reuse. Only active projects (`backlog`, `planned`, `in_progress`) should participate in shortname deduplication.

### Root cause

In `server/src/services/projects.ts`, both the `create` and `update` methods query **all** projects in the company for shortname collision checking:

```ts
const existingProjects = await db
  .select({ id: projects.id, name: projects.name })
  .from(projects)
  .where(eq(projects.companyId, existingProject.companyId));
```

This includes cancelled/completed projects, so their names are permanently "reserved."

### Fix

Filter out terminal-status projects from the dedup query:

```ts
.where(and(
  eq(projects.companyId, existingProject.companyId),
  notInArray(projects.status, ["cancelled", "completed"])
))
```

This applies to both `create` (line ~373) and `update` (line ~414).

### Impact

- Agents cannot programmatically clean up duplicate project names
- The CEO agent in our Bollinger instance hit this exact bug trying to consolidate "Growth Research 3" → "Growth Research" after cancelling duplicates

### PR DESCRIPTION
**Pull request by StartupBros**
## Summary
- ignore completed and cancelled projects when resolving project shortname collisions
- keep active projects participating in deduplication for both create and rename flows
- add regression coverage for terminal-vs-active project status handling

## Testing
- pnpm vitest run src/__tests__/project-shortname-resolution.test.ts
- pnpm typecheck

Fixes #553

### CODE REVIEW

**Review by mvanhorn — COMMENTED** [review]
## Summary

Clean fix for #553. The approach of filtering terminal projects at both the DB query level and in the pure function is solid defensive programming, and the tests cover the key scenarios well.

Two concerns worth addressing before merge — one correctness edge case and one maintainability nit. Also note that PR #562 fixes the same issue with a slightly different approach (uses typed `PROJECT_STATUSES` from `@paperclipai/shared`).

## Recommendation

A few changes needed (see inline comments).

---
*Review assisted by [Claude Code](https://claude.com/claude-code)*

File: server/src/services/projects.ts [hunk 1]
```diff
@@ -371,9 +375,14 @@ export function projectService(db: Db) {
       }
 
       const existingProjects = await db
-        .select({ id: projects.id, name: projects.name })
+        .select({ id: projects.id, name: projects.name, status: projects.status })
         .from(projects)
-        .where(eq(projects.companyId, companyId));
+        .where(
+          and(
+            eq(projects.companyId, companyId),
```
Comment (mvanhorn) [inline server/src/services/projects.ts]: **Potential NULL status exclusion**

In SQL, `status NOT IN ('completed', 'cancelled')` returns NULL (not TRUE) when `status` is NULL. If any projects have a NULL status column value, they'd be silently excluded from the dedup check - meaning you could create a name that collides with a NULL-status project.

The pure function handles this correctly with `!project.status || ...`, but those rows never reach it because the DB query already filters them out.

Consider wrapping with an `or(isNull(...))`:

```suggestion
        .where(
          and(
            eq(projects.companyId, companyId),
            or(
              isNull(projects.status),
              notInArray(projects.status, ["completed", "cancelled"]),
            ),
          ),
        );
```

(Same fix needed in the update path below.)

File: server/src/services/projects.ts [hunk 2]
```diff
@@ -371,9 +375,14 @@ export function projectService(db: Db) {
...
+            notInArray(projects.status, ["completed", "cancelled"]),
```
Comment (mvanhorn) [inline server/src/services/projects.ts]: **Hardcoded status strings vs constant**

`TERMINAL_PROJECT_STATUSES` is defined as a `Set` above, but these `notInArray` calls use hardcoded `["completed", "cancelled"]`. Consider defining the constant as an array (since `notInArray` needs an array anyway) and reusing it here:

```suggestion
            notInArray(projects.status, Array.from(TERMINAL_PROJECT_STATUSES)),
```

Or define `TERMINAL_PROJECT_STATUSES` as `["completed", "cancelled"] as const` instead of a `Set`, which would also give you better type inference.

**Comment by StartupBros** [comment]
Re-verified this against current `master` and pushed a clean fresh branch from latest upstream: `StartupBros:oss/fix-554-shortname-dedup`.

Current clean diff is still just:
- `server/src/services/projects.ts`
- `server/src/__tests__/project-shortname-resolution.test.ts`

Targeted test passes locally:
- `cd server && pnpm vitest run src/__tests__/project-shortname-resolution.test.ts`

Happy to open a fresh replacement PR from that branch if that is easier than merging this one.

**Comment by mvanhorn** [comment]
@StartupBros picking this back up because the underlying bug is still live, not just this PR.

Issue #553 is still open, and `TERMINAL_PROJECT_STATUSES` does not appear anywhere in master, so the fix never landed by another route. Three PRs took a run at it and all three are closed: this one, plus #561 and #562 from @TwilightSeibei. Whatever closed them, the rename is still blocked by cancelled and completed projects.

Your re-verified `oss/fix-554-shortname-dedup` branch therefore still applies. I only have read access here so I cannot reopen or merge, but for whoever can: the diff is two files, `server/src/services/projects.ts` and its test, and my review from March still stands.

If a fresh PR is easier than reopening this one, that seems worth doing rather than leaving #553 open with three closed fixes behind it.
